### PR TITLE
Allow instances to choose available units for products/variants

### LINKED ISSUE #6082: Allow instances to choose what available units for products/variants
## What is the problem we are solving 
With #5888 two new options (pounds and ounces) were introduced for product/variant units. The dropdown is now pretty big and could get bigger if we wanted to further support other units (many people in the US sell fruit by the pint or the quart), making it even more unwieldy to wade through a long list of options. 

This would allow shops to ignore units that they never use. 

Going further, we could also allow instances to choose the default available units for new shops on their instance (metric units only for most of the world, for example). 

## Success factors = expected outcome 
Happier admin users, easier product/variant creation

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Closes #6082 

After merging #5888 it came to light that some countries legally require products to only be sold in certain units. This PR allows an instance manager to specify which units appear during product creation/editing, allowing them to exclude units that are not allowed or just never used. 

The default units are metric, so new instances will likely never have to adjust this setting. 

We could make the settings UI a bit nicer and separate metric units, but I held off on putting too much time into that because it seems like there's some urgency to get this released. 

Screen Shot 2020-10-06 at 5 18 21 PM

#### What should we test?
Updating the available units in Configuration, going to the products tab and confirming that the correct units appear. 

#### Release notes
Added a setting to allow instance managers to select which units (kg, lb, etc) are available for products. 

Changelog Category: Added

#### Discourse thread
https://community.openfoodnetwork.org/t/allow-enterprises-to-choose-what-available-units-for-products-variants/2060

### CODE REVIEW

**Comment by andrewpbrett** [comment]
I don't really understand the test failures 😃 I do think they are legitimate, I'm just not sure why they're happening... I don't have a great understanding of the injection methods, I looked at how we do it for some other preferences and modeled it off of that.

**Comment by andrewpbrett** [comment]
Okay, the angular tests are now passing. There are three failures but they are passing locally for me, so I think it's flaky tests and we'll have a green build if we rerun. I'm moving to Code Review 👍

**Comment by luisramos0** [comment]
There's something weird in the build related to this https://github.com/openfoodfoundation/openfoodnetwork/issues/6142
I dont understand why these specs are breaking in semaphore with this error and not locally...
I will post on the dev slack channel about it now.

**Comment by luisramos0** [comment]
We found the problem! The problem in the build is 100% unrelated to this PR and will be solved in this PR: https://github.com/openfoodfoundation/openfoodnetwork/pull/6128

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (luisramos0) [inline app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:39]: this method is now very complicated. I think we need to extract a part of it to a separate method to make it more simple.

File: app/helpers/admin/injection_helper.rb
Comment (luisramos0) [inline app/helpers/admin/injection_helper.rb:187]: I'd prefer to have this as an API endpoint of configurations and not as injected data in the dom. What do you think?

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (luisramos0) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:39]: we would have a preferences api endpoint where we could expose quite a few of them including this one :tada:

File: app/views/spree/admin/general_settings/edit.html.haml
Comment (luisramos0) [inline app/views/spree/admin/general_settings/edit.html.haml:93]: I think it would be better named as available_units and not selected_units

File: app/controllers/spree/admin/general_settings_controller.rb
Comment (luisramos0) [inline app/controllers/spree/admin/general_settings_controller.rb:26]: maybe merge would be a better name than combine here? I didnt understand this initial. Now I think it's merging params :-)

File: app/controllers/spree/admin/general_settings_controller.rb
Comment (luisramos0) [inline app/controllers/spree/admin/general_settings_controller.rb:29]: Is there a way we can do this without using regexps? Can we use nested params? Something like params[:available_units][:g] ?

If we end up with this code, I think this if clause is too complicated and needs to be extracted to a method with a good name. Maybe:

```ruby
def extract_unit_from_param_name(param_name)
   param_name.match(/available_units_(.*)/)&.captures&.first
end
```
and then:

```ruby
if value == "1" && unit = extract_unit_from_param_name(name)
```

Now I see it's an assignment in an if clause, I thought it was a ==. Is rubocop not complaining about this?
In this case I recommend:
```ruby
if value == "1"
   unit = extract_unit_from_param_name(name)
   available_units << unit if unit
end
```

**Review by luisramos0 — CHANGES_REQUESTED** [review]
all right, good effort Andy, it's a nice feature!

I think we should improve a few things in this PR. Let me know what you think about my comments.

Thanks a lot for one more PR!!!!

File: app/helpers/admin/injection_helper.rb
Comment (luisramos0) [reply app/helpers/admin/injection_helper.rb:187]: we would have a preferences api endpoint where we could expose quite a few of them including this one :tada:

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (andrewpbrett) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:39]: 2a286af136c9a9365854afe5501123b37dce19ae

File: app/controllers/spree/admin/general_settings_controller.rb
Comment (andrewpbrett) [reply app/controllers/spree/admin/general_settings_controller.rb:29]: Good thinking - I didn't think to do nested params since there's not an ActiveRecord model that's backing this up, and there's not an option for fields_for in the base_helper.rb.

But! We can have it *act* like nested params and avoid the regex for sure: a0d4082d01dded8b81d2224c39e81a5e6c1cba99

I left it as a separated method (renamed to `merge_available_units_params` even though it's now just a one-liner.

File: app/helpers/admin/injection_helper.rb
Comment (andrewpbrett) [reply app/helpers/admin/injection_helper.rb:187]: Let me think about that some more... :) I can see both sides. On the one hand it's not so elegant to inject a string into the DOM in a bunch of places like we're doing now. And we will want to make this an API endpoint eventually anyway (since we want to make pretty much everything an API endpoint, right?)

I think my hesitation in having the front end call it as an endpoint is that it means an additional https call to the server, which 1) slows things down and 2) could make the code more complicated, since the methods that depend on knowing the value of `available_units` now have to wait for that request to return... but maybe we handle similar situations elsewhere and have a good pattern to follow for it. 

I think this tension will come up a lot, between having the front end javascript strictly use the API vs. having some data already available to it in the DOM. 

I'm going to sleep on it for now :)

Comment (luisramos0) [reply app/helpers/admin/injection_helper.rb:187]: yes, you comment is spot on.
My opinion is very simple: we must use the API approach in 95% of the cases if we ever want to have an useful API. Otherwise another 2/3 years will pass and not much api evolution will happen, like what happened in the last 2 years I have been around.

File: app/controllers/spree/admin/general_settings_controller.rb
Comment (luisramos0) [inline app/controllers/spree/admin/general_settings_controller.rb]: nice! I am very happy I asked, this is much better looking :clap:

Comment (luisramos0) [reply app/controllers/spree/admin/general_settings_controller.rb]: I like the fact that this is in a separate method :ok_hand:  the name of the method explains what's happening, much better than a comment.

**Review by luisramos0 — COMMENTED** [review]
Awesome, thanks Andy!

I am sorry but I am still not happy with the js code.

**Comment by andrewpbrett** [comment]
No need to be sorry about a copy paste detector, that's a useful thing to have. I agree it's still not very elegant and could be simplified... let me take another pass at it. This is where pairing would be useful :)

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (andrewpbrett) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:62]: It's because inside the scope of the function passed to `filter`, `@units` isn't available.... I think your suggestion of using a parameter will take care of it though.

File: app/helpers/admin/injection_helper.rb
Comment (andrewpbrett) [reply app/helpers/admin/injection_helper.rb:187]: Yes, I like keeping it disciplined. And I think this case is not so exceptional that it should be in the 5% of times when we don't use the API; it'll be a necessary endpoint if some client were trying to recreate the front end using only the API, so I think we should create it and eat our own dogfood.

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (luisramos0) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:62]: in variantUnitOptions:
```ruby
available = availableUnits.split(",")
for scale in @unitScales(unit_type, available)
```

and then rewrite unitScales with something  like this;:
```ruby
@unitScales: (unitType, availableUnits = nil) ->
    scales = Object.keys(@units[unitType])
    if availableUnits
         scales = scales.filter (scale) ->
           availableUnits.includes(units[unitType][scale]['name'])

   (parseFloat(scale) for scale in scales).sort (a, b) ->
       a - b
```

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (andrewpbrett) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:65]: 👍 a5c5aec33a98f3b1b453cc720ef6975ce702413a

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (andrewpbrett) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:59]: I still had to do `units = @units` but otherwise this works great: a5c5aec33a98f3b1b453cc720ef6975ce702413a

Comment (luisramos0) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:59]: great!
can you pls add a comment for that trick, I still dont understand why you need it :-)
I am still thinking, why is this not just: 
availableUnits.includes(@getUnitName: (scale, unitType))

Comment (andrewpbrett) [reply app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:59]: The trick was because inside the scope of the anonymous functions passed to `filter`, `@units` wasn't defined. I'm rereading it now and we could have just called `VariantUnitManager.units`, but instead, because of what you pointed out, we can use `VariantUnitManager.getUnitName` instead, and no more need for the trick. Thanks 🎉

File: app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee
Comment (luisramos0) [inline app/assets/javascripts/admin/products/services/variant_unit_manager.js.coffee:61]: I'd break this in two lines. One dot per line or one method call per line are good rules I think.

```ruby
unitName = VariantUnitManager.getUnitName(scale, unitType)
availableUnits.includes(unitName)
```

This is just my opinion.

**Review by luisramos0 — APPROVED** [review]
Nice, well done, looks good to me!
Thanks Andy!

**Comment by luisramos0** [comment]
This is now waiting for the second review :+1:

**Review by Matt-Yorkley — APPROVED** [review]
Awesome :muscle:

**Comment by filipefurtad0** [comment]
Hi @andrewpbrett ,

This PR had a second run of the build, just to check if these specs would pass:

```
rspec ./spec/features/consumer/shopping/checkout_stripe_spec.rb:120 # Check out with Stripe using Stripe SCA with guest checkout when the card is rejected shows an error message from the Stripe response
rspec ./spec/features/consumer/shopping/checkout_stripe_spec.rb:169 # Check out with Stripe using Stripe SCA with guest checkout when the card needs extra SCA authorization and the authorization fails shows an error message from the Stripe response
rspec ./spec/features/consumer/shopping/checkout_stripe_spec.rb:187 # Check out with Stripe using Stripe SCA with guest checkout with multiple payment attempts; one failed and one succeeded records failed payment attempt and allows order completion

```

But from rebuilding this some conflicts arose - please see above. Would it be possible to address these, before proceeding with testing? Thanks!

**Comment by sigmundpetersen** [comment]
I think a rebase would be good to try first. There are also some conflicts that need to be sorted before merging.

**Comment by luisramos0** [comment]
yeah, conflicts should always be addressed before testing :+1:

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

Thanks for addressing this. After staging this PR, I found the unit check-boxes under Configuration -> General Settings menu (as super admin):

image


Verified that:

- the default setting is metric (as shown above)
- the changes on these **settings are reflected** in/when:
i) products page
ii) creating a new product, ex. below:

image

- the changes on these **settings are not reflected** in:

iii) weight calculators (for enterprise or shipping fees)

image

Kg option is available, despite having previously opted out.

iv) bulk order management

Weight/volume is kept in metric (g), as before:

image

v) regression test: opting-out from all units has not influence on checking out with previously created products/fees.

All good.

I can't seem to think of any other tests for this.
It looks good to be merged, thanks Andy!

**Pull request merged**
